Create TableStats interface on Admin API

mrtracy opened the pull request:
Commit adds TableStats interface, which returns computationally expensive
information about a specific table.

This following operations are considered expensive:
- Scan of Meta2 keys for the table
- Fan-out query to all nodes which contain data for the table

The "Fan-out" query calls SpanStats on each node, returning aggregated
statistics for the target table on that node. The results of all SpanStats
queries are reduced into the result of TableStats.

Owing to its high level and complexity, this method is tested via a new
acceptance test.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8018)

<!-- Reviewable:end -->

Commit:
Create TableStats interface on Admin API

Commit adds TableStats interface, which returns computationally expensive
information about a specific table.

This following operations are considered expensive:
+ Scan of Meta2 keys for the table
+ Fan-out query to all nodes which contain data for the table

The "Fan-out" query calls SpanStats on each node, returning aggregated
statistics for the target table on that node. The results of all SpanStats
queries are reduced into the result of TableStats.

tamird commented:
Could you test this with a TestCluster instead of an acceptance test?

---

Reviewed 8 of 9 files at r1.
Review status: 8 of 9 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[server/admin.go, line 447 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZNlxuV9az5uoHkcv0:-KNZNlxuV9az5uoHkcv1:1549940641) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L447)):*
> ```Go
> 	*serverpb.TableStatsResponse, error,
> ) {
> 	escDbName := parser.Name(req.Database).String()
> ```

s/Db/DB/

---

*[server/admin.go, line 475 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZOQxqjo93A3aHvA8Q:-KNZOQxqjo93A3aHvA8R:-1017599519) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L475)):*
> ```Go
> 	// meta2 keys for the range.
> 	var resp *roachpb.ScanResponse
> 	{
> ```

is this scope necessary? i'd prefer `resp := ...` below.

---

*[server/admin.go, line 476 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZOO8d5-IK-atwFJy1:-KNZOO8d5-IK-atwFJy2:1609255290) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L476)):*
> ```Go
> 	var resp *roachpb.ScanResponse
> 	{
> 		// Directly scan metadata keys for the table.
> ```

I think this needs more words explaining why client.Batch can't be used.

---

*[server/admin.go, line 494 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZOWv79Uhuytl_Foay:-KNZOWv79Uhuytl_Foaz:2089831176) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L494)):*
> ```Go
> 	// Extract a list of node IDs from the response.
> 	nodeIDs := make(map[roachpb.NodeID]struct{})
> 	{
> ```

this scope buys you exactly nothing, since the `for` below is already a scope.

---

*[server/admin.go, line 511 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZOat2pbhJJL_9FobN:-KNZOat2pbhJJL_9FobO:1048840742) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L511)):*
> ```Go
> 		RangeCount: resp.Count(),
> 	}
> 	{
> ```

again, why?

---

*[server/admin.go, line 525 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZOh9AHQHOeG34h10c:-KNZOh9AHQHOeG34h10d:761943954) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L525)):*
> ```Go
> 		var (
> 			fanoutDeadline  time.Time
> 			ctxWithDeadline context.Context
> ```

why are you declaring `ctxWithDeadline` and `cancel` up here? they can just as easily be `:=`'d below.

---

*[server/admin.go, line 535 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZPGE-UqFdUkjdw_Zl:-KNZPGE-UqFdUkjdw_Zm:1701403504) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L535)):*
> ```Go
> 			fanoutDeadline = now.Add(base.NetworkTimeout)
> 		case now.Add(base.NetworkTimeout).After(deadline):
> 			// Deadline will arrive in less than base.NetworkTimeout. In this
> ```

what does it matter that the deadline will arrive before base.NetworkTimeout? We only need to reserve a tiny amount of time here.

```
var (
  ctxWithDeadline context.Context
  cancel context.CancelFunc
)
if deadline, ok := ctx.Deadline(); ok {
  ctxWithDeadline, cancel = ctx.WithDeadline(ctx, deadline.Sub(100*time.Millisecond))
} else {
  ctxWithDeadline, cancel = ctx.WithTimeout(ctx, base.NetworkTimeout)
}
```

---

*[server/admin.go, line 578 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQE0q4bHf9eLODF8k:-KNZQE0q4bHf9eLODF8l:1407309048) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L578)):*
> ```Go
> 		}
> 		remainingResponses := len(nodeIDs)
> 		for remainingResponses > 0 {
> ```

you can put all the mutation of `remainingResponses` into this `for` line:
```
for remaining := len(nodeIDs); remaining > 0; remaining--- {
```

---

*[server/admin.go, line 583 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQQLuIuvfoqPVqZOy:-KNZQQLuIuvfoqPVqZOz:2033224150) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L583)):*
> ```Go
> 				// For nodes which returned an error, note that the node's data
> 				// is missing. For successful calls, aggregate statistics.
> 				if resp.err != nil {
> ```

don't you want to keep the error?

---

*[server/serverpb/admin.proto, line 152 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQYlncul-g7dK_lzR:-KNZQYlncul-g7dK_lzS:-1819534690) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L152)):*
> ```protobuf
>   // database is the database that contains the table we're interested in.
>   string database = 1;
> ```

well, are we doing spaces or not?

---

*[server/serverpb/admin.proto, line 166 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQdP1Yvxa0n7ffB_9:-KNZQdP2GXGT920Xbnsv:1750451392) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L166)):*
> ```protobuf
>   // this query.
>   int64 replica_count = 2;
>   // node_count is the number of nodes which should contain data for this table.
> ```

should? explanation below is good, add it here.

---

*[server/serverpb/admin.proto, line 168 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQgFR_drrbv-J2oFc:-KNZQgFR_drrbv-J2oFd:831921228) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L168)):*
> ```protobuf
>   // node_count is the number of nodes which should contain data for this table.
>   int64 node_count = 3;
>   // mvcc_stats is the summation of MVCCStats for all replicas of this table
> ```

replicas or ranges?

---

*[server/serverpb/admin.proto, line 173 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQVHubZr7gZsJH7NS:-KNZQVHubZr7gZsJH7NT:84535935) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L173)):*
> ```protobuf
>   // A list of nodes which should contain data for this table (according to
>   // cluster metadata), but could not be contacted during this request.
>   repeated string missing_nodes = 5;
> ```

I think you probably want to store the error as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-:-KNZNfvYIx7U-TJr5rtJ:-591668700)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Agree with @tamird that it would be preferable to use `TestCluster` instead of an acceptance test. Acceptance tests should mainly be used for testing that requires a particular build environment (e.g. the various language driver tests) or runtime environment (i.e. `TestPartition*` which manipulates iptables). My thinking on this has evolved with the recent addition of `TestCluster`. We should probably document this somehow in the `acceptance` package.

---

Review status: 8 of 9 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-:-KNZ_qAQCbG6AKzy7JR1:-1459194925)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 8 of 9 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[server/admin.go, line 447 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZNlxuV9az5uoHkcv0:-KNbchmtijmNNIJTdG7U:-1061391158) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L447)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/Db/DB/
</details>
Is escaping the right thing to be doing here? This will convert a db name of `foo bar` into `"foo bar"` which seems wrong to me.

---

*[server/admin.go, line 548 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNbapmNm7kcfZBVndPf:-KNbapmOmvzWuQpRoBNE:1339936061) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L548)):*
> ```Go
> 			// this server.
> 			fanoutDeadline = deadline.Add(-1 * base.NetworkTimeout)
> 		}
> ```

Are all of these cases being tested?

Note that `context.WithDeadline` already takes care of making sure the new deadline is less than the deadline of the parent context.

---

*[server/admin.go, line 550 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNbbY0lZncke_En-BFW:-KNbbY0lZncke_En-BFX:1010628279) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L550)):*
> ```Go
> 		}
> 
> 		ctxWithDeadline, cancel = context.WithDeadline(context.Background(), fanoutDeadline)
> ```

I think this should be `context.WithDeadline(ctx, fanoutDeadline)`. That ties cancellation of `ctx` to cancellation of `ctxWithDeadline`. 

---

*[server/admin.go, line 555 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNbcFz7wckbhTvn3igh:-KNbcFz7wckbhTvn3igi:-2023843490) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L555)):*
> ```Go
> 		// Send a SpanStats query to each node.
> 		for nodeID := range nodeIDs {
> 			id := nodeID
> ```

I have seen this more idiomatically expressed as `nodeID := nodeID`. That prevents accidental use of the wrong `nodeID` in the closure below.

---

*[server/serverpb/admin.proto, line 170 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNba47w0DEwNqItNhcl:-KNba47yA5XuxFxJyJqD:-1495477084) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L170)):*
> ```protobuf
>   // mvcc_stats is the summation of MVCCStats for all replicas of this table
>   // across the cluster.
>   cockroach.storage.engine.enginepb.MVCCStats mvcc_stats = 4 [(gogoproto.nullable) = false];
> ```

`s/mvcc_stats/stats/g` or give this field a custom name so that it is generated as `MVCCStats` (not `MvccStats`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[server/admin.go, line 530 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNbgJYNhPwR9wayU19j:-KNbgJYNhPwR9wayU19k:-697004645) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L530)):*
> ```Go
> 		now := s.server.clock.PhysicalTime()
> 		deadline, ok := ctx.Deadline()
> 		switch {
> ```

This seems overly elaborate. What do you lose if you completely ignore what deadline the timeout already has and just wrap it with a `base.NetworkTimeout` deadline?

---

*[server/serverpb/admin.proto, line 164 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNbguuzCSVaDxvifNEA:-KNbguv-6eM8FE5Ia-3x:-1007796243) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L164)):*
> ```protobuf
>   int64 range_count = 1;
>   // replica_count is the number of active replicas which were found during
>   // this query.
> ```

From this description, I'm not sure what is being returned here. Can you elaborate more? If the table spans two Ranges, and one range is unavailable (2/3 down) and the other one is overreplicated (4/3) but two of the Replicas are dormant due to inactivity, do I get 2+2=4?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[server/admin.go, line 555 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNbcFz7wckbhTvn3igh:-KNbneti7AnRTwXtwgiE:676269983) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L555)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I have seen this more idiomatically expressed as `nodeID := nodeID`. That prevents accidental use of the wrong `nodeID` in the closure below.
</details>
:+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Code comments addressed. This now depends on the earlier commit to modify TestCluster, which is being reviewed in a separate PR.  The earlier PR will be merged first.

---

Review status: 0 of 17 files reviewed at latest revision, 19 unresolved discussions, some commit checks pending.

---

*[server/admin.go, line 447 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZNlxuV9az5uoHkcv0:-KNc8IRbXbCFEyCyPEww:1683538059) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L447)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Is escaping the right thing to be doing here? This will convert a db name of `foo bar` into `"foo bar"` which seems wrong to me.
</details>
This escaping is what is being used in the existing "TableDetails" API; these values just feed directly into GetTableSpan, which normalizes the names and looks up descriptor IDs.   Is this escaping incorrect for that purpose?

---

*[server/admin.go, line 475 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZOQxqjo93A3aHvA8Q:-KNc8gO4uOwKj3tGBKQe:1183282066) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L475)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this scope necessary? i'd prefer `resp := ...` below.
</details>
This is the style already being used inside of this file.  I do find it to be very clean; it breaks up a longer function into subsections, without needing to break it up into sub-functions.

That said, where it was being used earlier, the individual scopes were essentially unrelated. In this function, each section is required to compute the next; after reconsidering, i've removed these additional scopes in this function

---

*[server/admin.go, line 476 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZOO8d5-IK-atwFJy1:-KNcSSIvS0gGryxXGgBq:-1676258102) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L476)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think this needs more words explaining why client.Batch can't be used.
</details>
I was emulating code in another method of this file which does an inconsistent read.  I want a consistent read here, which means that it appropriate to use the server's KV client instead of the distsender directly.  Modified.

---

*[server/admin.go, line 494 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZOWv79Uhuytl_Foay:-KNidNaFVTNAPqK-8bNJ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L494)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this scope buys you exactly nothing, since the `for` below is already a scope.
</details>
Done.

---

*[server/admin.go, line 511 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZOat2pbhJJL_9FobN:-KNidO6nrMmdH6_tGlsl:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L511)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> again, why?
</details>
Done.

---

*[server/admin.go, line 525 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZOh9AHQHOeG34h10c:-KNidQgZmHRjg5mtoZN8:1881417684) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L525)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why are you declaring `ctxWithDeadline` and `cancel` up here? they can just as easily be `:=`'d below.
</details>
See my response to toby.

---

*[server/admin.go, line 530 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNbgJYNhPwR9wayU19j:-KNidSNUNyKrRLVz22i_:-75743754) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L530)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This seems overly elaborate. What do you lose if you completely ignore what deadline the timeout already has and just wrap it with a `base.NetworkTimeout` deadline?
</details>
I've chosen this; I put much of this in when I was diagnosing some timeout errors in a test, but this turned out not to be the cause. I am now using base.NetworkTimeout instead.

---

*[server/admin.go, line 535 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZPGE-UqFdUkjdw_Zl:-KNidfi4GEQ2MFmiX-dm:-1582418045) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L535)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what does it matter that the deadline will arrive before base.NetworkTimeout? We only need to reserve a tiny amount of time here.
> 
> ```
> var (
>   ctxWithDeadline context.Context
>   cancel context.CancelFunc
> )
> if deadline, ok := ctx.Deadline(); ok {
>   ctxWithDeadline, cancel = ctx.WithDeadline(ctx, deadline.Sub(100*time.Millisecond))
> } else {
>   ctxWithDeadline, cancel = ctx.WithTimeout(ctx, base.NetworkTimeout)
> }
> ```
</details>
See above.

---

*[server/admin.go, line 548 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNbapmNm7kcfZBVndPf:-KNidgdrtkLhBaRDmAMs:-1582418045) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L548)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Are all of these cases being tested?
> 
> Note that `context.WithDeadline` already takes care of making sure the new deadline is less than the deadline of the parent context.
</details>
See above.

---

*[server/admin.go, line 550 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNbbY0lZncke_En-BFW:-KNcbH4BV6kaRdFuVn0s:265855168) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L550)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I think this should be `context.WithDeadline(ctx, fanoutDeadline)`. That ties cancellation of `ctx` to cancellation of `ctxWithDeadline`. 
</details>
Sorry, we were diagnosing a weird issue yesterday (caused by the loop invariant below) and did not change this back. Fixed.

---

*[server/admin.go, line 555 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNbcFz7wckbhTvn3igh:-KNcb9D03ZZavvlnWikN:310136278) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L555)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> :+1:
</details>
I didn't think that would be allowed due to shadowing rules. That would have saved me some trouble, I overthought things.

---

*[server/admin.go, line 578 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQE0q4bHf9eLODF8k:-KNieUUgdvCwhX3EUo8j:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L578)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you can put all the mutation of `remainingResponses` into this `for` line:
> ```
> for remaining := len(nodeIDs); remaining > 0; remaining--- {
> ```
</details>
Done.

---

*[server/admin.go, line 583 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQQLuIuvfoqPVqZOy:-KNih-YFskUpGzZuIXoE:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L583)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> don't you want to keep the error?
</details>
Done.

---

*[server/serverpb/admin.proto, line 152 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQYlncul-g7dK_lzR:-KNieZzT3Byy4rYcE6vp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L152)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> well, are we doing spaces or not?
</details>
Done.

---

*[server/serverpb/admin.proto, line 164 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNbguuzCSVaDxvifNEA:-KNigagfPVmOQc1LHWVD:1140256980) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L164)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> From this description, I'm not sure what is being returned here. Can you elaborate more? If the table spans two Ranges, and one range is unavailable (2/3 down) and the other one is overreplicated (4/3) but two of the Replicas are dormant due to inactivity, do I get 2+2=4?
</details>
Clarified a bit.

---

*[server/serverpb/admin.proto, line 166 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQdP1Yvxa0n7ffB_9:-KNigaNpjqWJosCX2xz8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L166)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> should? explanation below is good, add it here.
</details>
Done.

---

*[server/serverpb/admin.proto, line 168 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQgFR_drrbv-J2oFc:-KNies3z0bHbux_P00E6:-1325248146) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L168)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> replicas or ranges?
</details>
replicas.

---

*[server/serverpb/admin.proto, line 170 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNba47w0DEwNqItNhcl:-KNigwsXOnwEXGA3tpB5:-1085629579) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L170)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/mvcc_stats/stats/g` or give this field a custom name so that it is generated as `MVCCStats` (not `MvccStats`).
</details>
Renamed to stats.

---

*[server/serverpb/admin.proto, line 173 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZQVHubZr7gZsJH7NS:-KNigydeD7vOGeywd1cG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/serverpb/admin.proto#L173)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think you probably want to store the error as well.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-:-KNihFOBGcoCBRD0Amys:-313148476)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 17 files reviewed at latest revision, 22 unresolved discussions, some commit checks failed.

---

*[server/admin.go, line 447 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZNlxuV9az5uoHkcv0:-KNiqQUZ5L-hMDUgWtns:977163181) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L447)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> This escaping is what is being used in the existing "TableDetails" API; these values just feed directly into GetTableSpan, which normalizes the names and looks up descriptor IDs.   Is this escaping incorrect for that purpose?
</details>
Yes, these values are getting passed to `GetTableSpan` which, as far as I can tell, doesn't want escaped names. You can verify this. Change your test to use a database such as `foo bar` and a table name which contains a space as well. Escaping like this is necessary when we're going to parse the name later, but `GetTableSpan` isn't parsing (or lexing) the supplied names.

PS I noticed that escaping is being performed elsewhere in this file, but that is done before constructing a SQL statement that will be executed. You're not constructing a SQL statement here.

---

*[server/admin_cluster_test.go, line 51 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNisC3DFofU5t01VmJ-:-KNisC3DFofU5t01VmJ0:-1810687543) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L51)):*
> ```Go
> 	client.Timeout = base.NetworkTimeout * 3
> 
> 	if err != nil {
> ```

Shouldn't you perform this check before assigning to `client.Timeout`?

---

*[server/admin_cluster_test.go, line 56 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNisIjSS_FYuP2c1y8_:-KNisIjSS_FYuP2c1y8a:1339953907) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L56)):*
> ```Go
> 
> 	// Make a single table and insert some data.
> 	if _, err := db.Exec("CREATE DATABASE test"); err != nil {
> ```

Per my other comment, try changing the database name to `"test test"`. Similarly, try naming the table `"foo foo"`. 

---

*[server/admin_cluster_test.go, line 128 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNitFXRgRk9lEUd7Knq:-KNitFXRgRk9lEUd7Knr:-886458555) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L128)):*
> ```Go
> 	// that fan-out goroutines are not leaked by the server if the request is
> 	// cancelled.
> 	client.Timeout = 50 * time.Millisecond
> ```

50ms is low, but still seems possible for the query to succeed in that time frame. Any reason to not make the absurdly low (i.e. 1ns)?

---

*[server/admin_cluster_test.go, line 130 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNitRx9TAel0NrglGXA:-KNitRx9TAel0NrglGXB:606917489) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L130)):*
> ```Go
> 	client.Timeout = 50 * time.Millisecond
> 	if err := util.GetJSON(client, url, &tsResponse); err == nil {
> 		t.Fatal("Expected low-timeout request to fail, but it succeeded.")
> ```

`s/Expected/expected/g` (everywhere in this function) to match convention. 

---

*[server/serverpb/admin.proto, line 177 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNirSsLsfGzZBS0Ikyj:-KNirSsLsfGzZBS0Ikyk:-267758975) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/serverpb/admin.proto#L177)):*
> ```protobuf
>   message MissingNode {
>     // The ID of the missing node.
>     string node_id = 1;
> ```

Shouldn't this be an `int32` with a custom type of `roachpb.NodeID`. Also, you need a custom name here so that this field is generated as `NodeID` instead of `NodeId`. There are examples of doing this in the `roachpb` protos.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Feedback addressed, but part of the test depends on #8206

---

Review status: 0 of 14 files reviewed at latest revision, 22 unresolved discussions, some commit checks pending.

---

*[server/admin.go, line 447 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNZNlxuV9az5uoHkcv0:-KO7X5WCiEkY9Mw5dMiK:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc56fa3208e20b49b4e7731dcfe024e06c8b95d1/server/admin.go#L447)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yes, these values are getting passed to `GetTableSpan` which, as far as I can tell, doesn't want escaped names. You can verify this. Change your test to use a database such as `foo bar` and a table name which contains a space as well. Escaping like this is necessary when we're going to parse the name later, but `GetTableSpan` isn't parsing (or lexing) the supplied names.
> 
> PS I noticed that escaping is being performed elsewhere in this file, but that is done before constructing a SQL statement that will be executed. You're not constructing a SQL statement here.
</details>
Done.

---

*[server/admin_cluster_test.go, line 51 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNisC3DFofU5t01VmJ-:-KO7X4BgmdSbrMi7nlRp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L51)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Shouldn't you perform this check before assigning to `client.Timeout`?
</details>
Done.

---

*[server/admin_cluster_test.go, line 56 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNisIjSS_FYuP2c1y8_:-KO7X3svrqJEGLMAr5M3:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L56)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Per my other comment, try changing the database name to `"test test"`. Similarly, try naming the table `"foo foo"`. 
</details>
Done.

---

*[server/admin_cluster_test.go, line 128 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNitFXRgRk9lEUd7Knq:-KO7_N0rW03b2DOLsa4g:-195430105) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L128)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> 50ms is low, but still seems possible for the query to succeed in that time frame. Any reason to not make the absurdly low (i.e. 1ns)?
</details>
Apparently, it can now succeed even with a 1ns timeout.  I've switched it to use 1ns, but we might simply need to *not care* if it succeeds or not; either that, or add a test hook to the server which guarantees a timeout.  I'm not sure if that's worth it.

This check is also dependent on #8206

---

*[server/admin_cluster_test.go, line 130 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNitRx9TAel0NrglGXA:-KO7Vt2SY7NYZCXA2FAg:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L130)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/Expected/expected/g` (everywhere in this function) to match convention. 
</details>
Done.

---

*[server/serverpb/admin.proto, line 177 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNirSsLsfGzZBS0Ikyj:-KO7XBpmUxx4i_J3HBTI:1689868532) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/serverpb/admin.proto#L177)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Shouldn't this be an `int32` with a custom type of `roachpb.NodeID`. Also, you need a custom name here so that this field is generated as `NodeID` instead of `NodeId`. There are examples of doing this in the `roachpb` protos.
</details>
If it should be int32, we need to change the rest of the admin api - node ID is being represented as string everywhere.  @cuongdo, was there a reason for using strings instead of int32?

Added the custom name.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-:-KO7_FKnBOK_sr_7IDfu:-2035299474)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 12 of 14 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[server/admin.go, line 548 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KO8AHDRnNQXSZYpRprH:-KO8Cmayuf9pZLssA62l:-917301825) ([raw file](https://github.com/cockroachdb/cockroach/blob/600c50f1730bd9b373da381608c2d1acbb65fd27/server/admin.go#L548)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you want to use ctxWithTimeout here. if `ctx`'s timeout is very large or infinite, this can deadlock. in fact, there's no reason not to simply shadow `ctx` with the result of `context.WithTimeout`.
> 
> it's gonna be hard to test because it can only happen during shutdown where the stopper is draining.
</details>
on second thought, it probably can't happen, but still no reason to complicate this with multiple contexts in scope.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 11 of 14 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[server/admin.go, line 502 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KO89wFHAtd77wFldDE7:-KO98B4jIXLvX4CO7A2Q:-1700829672) ([raw file](https://github.com/cockroachdb/cockroach/blob/600c50f1730bd9b373da381608c2d1acbb65fd27/server/admin.go#L502)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why does each node need its own context? you can use a single one for them all.
</details>
Wasn't sure about the cancel function, but i'll move it up.

---

*[server/admin.go, line 548 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KO8AHDRnNQXSZYpRprH:-KO99-2Gbi-wmokbUX3D:-948564508) ([raw file](https://github.com/cockroachdb/cockroach/blob/600c50f1730bd9b373da381608c2d1acbb65fd27/server/admin.go#L548)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> on second thought, it probably can't happen, but still no reason to complicate this with multiple contexts in scope.
</details>
This is exactly where I found the deadlock earlier; i was trying to test it by calling tablestats with a very small timeout in the test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 15 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[server/admin_cluster_test.go, line 128 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNitFXRgRk9lEUd7Knq:-KOAh12F6cz48st_nn-u:1204934924) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L128)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Apparently, it can now succeed even with a 1ns timeout.  I've switched it to use 1ns, but we might simply need to *not care* if it succeeds or not; either that, or add a test hook to the server which guarantees a timeout.  I'm not sure if that's worth it.
> 
> This check is also dependent on #8206
</details>
You'll need to add the extra goroutine to the leaktest whitelist as discussed in https://github.com/cockroachdb/cockroach/issues/8204#issuecomment-236826147

Also, this is still 50ns, let's go ahead and make it 1ns. Finally, we should get to the bottom of how on earth this is succeeding - it really shouldn't be.

---

*[server/admin_cluster_test.go, line 47 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAfMMnXFii2Il499kS:-KOAfMMo-7AVEmzLXJ6r:-1583348447) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L47)):*
> ```Go
> 	server0 := tc.Server(0)
> 
> 	// Create clients (SQL, Http) connected to server 0.
> ```

s/Http/HTTP/

---

*[server/admin_cluster_test.go, line 67 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAfXLVKuDH-grPbVUT:-KOAfXLVKuDH-grPbVUU:-442973565) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L67)):*
> ```Go
> 			id INT PRIMARY KEY,
> 			val STRING
> 		)`); err != nil {
> ```

nit: make the line containing the function call's closing paren line up with the line containing the opening paren. either:
```go
db.Exec(`CREATE TABLE ......
  ....
  ...)`,
)
```
or
```go
db.Exec(`
  CREATE TABLE ....
   ..)`,
)
```

Perhaps it's time we discussed broadly which style we prefer, since this has come up a few times recently.

---

*[server/admin_cluster_test.go, line 81 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAg6Z1c2w_2Cpi89wA:-KOAg6Z24--cjdxE2r-I:1892406992) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L81)):*
> ```Go
> 	var tsResponse serverpb.TableStatsResponse
> 
> 	// The new SQL table may not yet have split into its own range yet. Wait
> ```

this sentence contains the word "yet" twice.

---

*[server/admin_cluster_test.go, line 90 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAgE5d3A8dShu6_E-J:-KOAgE5d3A8dShu6_E-K:-1217705567) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L90)):*
> ```Go
> 			return errors.Errorf("Table range not yet separated.")
> 		}
> 		if tsResponse.NodeCount != 3 {
> ```

can you avoid this magic number? it's the parameter you passed to StartTestCluster.

---

*[server/admin_cluster_test.go, line 97 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAgKmhRSAIzE2ZITnd:-KOAgKmhRSAIzE2ZITne:1755550784) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L97)):*
> ```Go
> 		}
> 		// These two conditions *must* be true, given that the above
> 		// SucceedsSoon has succeeded.
> ```

Move these out of the retry loop. This comment is currently weird to read.

---

*[server/admin_cluster_test.go, line 114 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAgaOl_nsgH7j0uNNj:-KOAgaOl_nsgH7j0uNNk:-1550317518) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L114)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	if a, e := tsResponse.NodeCount, int64(3); a != e {
> ```

magic numbers here

by the way, can you just assert on the entire tsResponse structure instead of having 4 separate assertions? i'm not sure why the number of keys expected needs to be "at least".

---

*[server/status_test.go, line 599 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAhQdr2FF8oPBTlzGX:-KOAhQdsO8UFwmJ5yReF:-673758678) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/status_test.go#L599)):*
> ```Go
> 	}
> 
> 	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
> ```

why the timeout?

---

*[server/status_test.go, line 604 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAhMrrOKqOBbtVZyDI:-KOAhMrrOKqOBbtVZyDJ:991263756) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/status_test.go#L604)):*
> ```Go
> 	url := ts.ServingAddr()
> 	conn, err := rpcContext.GRPCDial(url)
> ```

nuke this newline

---

*[server/status_test.go, line 606 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAhKb6_s7xrylwvSYB:-KOAhKb6_s7xrylwvSYC:401546512) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/status_test.go#L606)):*
> ```Go
> 
> 	if err != nil {
> 		t.Fatalf("ERROR CONNECTING INITIALLY: %s", err)
> ```

what's with the caps? just `t.Fatal(err)` here.

---

*[server/status_test.go, line 614 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAhWi2memKJ3Je4t8X:-KOAhWi2memKJ3Je4t8Y:210006225) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/status_test.go#L614)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	if a, e := int(response.RangeCount), ExpectedInitialRangeCount(); a != e {
> ```

seems better to upcase ExpectedInitialRangeCount to int64 than to (potentially) downcast RangeCount to int32. There's a thread somewhere about supporting 32bit, so we need to start being careful about this.

---

*[server/status_test.go, line 619 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAhg93RlzLBonoQ3ef:-KOAhg94dbUjZvHTGVO1:-362946067) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/status_test.go#L619)):*
> ```Go
> 
> 	ts.Stopper().Stop()
> 	<-ts.Stopper().IsStopped()
> ```

What's up with this and the commented out code below? Why do you need to wait for IsStopped? Stop doesn't return until IsStopped is closed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[server/admin_cluster_test.go, line 128 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNitFXRgRk9lEUd7Knq:-KOAspSADZ1Meg3avViJ:1468467690) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L128)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> You'll need to add the extra goroutine to the leaktest whitelist as discussed in https://github.com/cockroachdb/cockroach/issues/8204#issuecomment-236826147
> 
> Also, this is still 50ns, let's go ahead and make it 1ns. Finally, we should get to the bottom of how on earth this is succeeding - it really shouldn't be.
</details>
Why shouldn't this be succeeding? He's explicitly trying to send a request and have it cancelled almost immediately.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[server/admin_cluster_test.go, line 128 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNitFXRgRk9lEUd7Knq:-KOAtXGPnf3pW0Op5lSU:1626251247) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L128)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Why shouldn't this be succeeding? He's explicitly trying to send a request and have it cancelled almost immediately.
</details>
Are you suggesting that cancellation is considered success? That's news to me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[server/admin_cluster_test.go, line 128 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNitFXRgRk9lEUd7Knq:-KOAukPijIT6tYIA4i37:-52697738) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L128)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Are you suggesting that cancellation is considered success? That's news to me.
</details>
Yes. I think Matt's trying to test that the cancellation propagates to the goroutines he forked off.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[server/admin_cluster_test.go, line 128 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNitFXRgRk9lEUd7Knq:-KOAv8dV0kR8ikGHYLKq:1347762852) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L128)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yes. I think Matt's trying to test that the cancellation propagates to the goroutines he forked off.
</details>
Yes, he is, and that's verified by leaktest. By "success" Matt means that GetJSON is returning a nil error here, and that's the surprising bit.

Why would the HTTP client consider a timed out request successful?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 11 of 15 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[server/admin_cluster_test.go, line 128 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KNitFXRgRk9lEUd7Knq:-KOBEQhLu0sW0X5inIdN:1832173370) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f6ae78f58b73e72e0fe58d4c7f71287e5484ba0/server/admin_cluster_test.go#L128)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Yes, he is, and that's verified by leaktest. By "success" Matt means that GetJSON is returning a nil error here, and that's the surprising bit.
> 
> Why would the HTTP client consider a timed out request successful?
</details>
Added to the goroutine whitelist.

---

*[server/admin_cluster_test.go, line 47 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAfMMnXFii2Il499kS:-KOBBzeS-Sl4K6XPNT6A:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L47)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/Http/HTTP/
</details>
Done.

---

*[server/admin_cluster_test.go, line 67 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAfXLVKuDH-grPbVUT:-KOBCwJX-wY8C6jzRcS6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L67)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: make the line containing the function call's closing paren line up with the line containing the opening paren. either:
> ```go
> db.Exec(`CREATE TABLE ......
>   ....
>   ...)`,
> )
> ```
> or
> ```go
> db.Exec(`
>   CREATE TABLE ....
>    ..)`,
> )
> ```
> 
> Perhaps it's time we discussed broadly which style we prefer, since this has come up a few times recently.
</details>
Done.

---

*[server/admin_cluster_test.go, line 81 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAg6Z1c2w_2Cpi89wA:-KOBD9OPc4e7sFK1ZvPE:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L81)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this sentence contains the word "yet" twice.
</details>
Done.

---

*[server/admin_cluster_test.go, line 90 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAgE5d3A8dShu6_E-J:-KOBDt7QOrlD0ugfh9Qo:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L90)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can you avoid this magic number? it's the parameter you passed to StartTestCluster.
</details>
Done.

---

*[server/admin_cluster_test.go, line 97 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAgKmhRSAIzE2ZITnd:-KOBDtjpTxlFSBH5a5eR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L97)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Move these out of the retry loop. This comment is currently weird to read.
</details>
Done.

---

*[server/admin_cluster_test.go, line 114 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAgaOl_nsgH7j0uNNj:-KOBDuRz0ur85Kqr3LGO:-1406555894) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/admin_cluster_test.go#L114)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> magic numbers here
> 
> by the way, can you just assert on the entire tsResponse structure instead of having 4 separate assertions? i'm not sure why the number of keys expected needs to be "at least".
</details>
"at least" because one of the replicas may be behind.

---

*[server/status_test.go, line 599 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAhQdr2FF8oPBTlzGX:-KOBG7H0eWaySxyPKKAz:1009862297) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/status_test.go#L599)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why the timeout?
</details>
Probably added during debugging, looks like this test needed some cleanup.

---

*[server/status_test.go, line 604 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAhMrrOKqOBbtVZyDI:-KOBH5U8WdGzlHg4ODO_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/status_test.go#L604)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nuke this newline
</details>
Done.

---

*[server/status_test.go, line 606 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAhKb6_s7xrylwvSYB:-KOBH5sYBDhchkeODT1S:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/status_test.go#L606)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's with the caps? just `t.Fatal(err)` here.
</details>
Done.

---

*[server/status_test.go, line 614 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAhWi2memKJ3Je4t8X:-KOBHQacyjpFJVx57nwk:53963764) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/status_test.go#L614)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> seems better to upcase ExpectedInitialRangeCount to int64 than to (potentially) downcast RangeCount to int32. There's a thread somewhere about supporting 32bit, so we need to start being careful about this.
</details>
RangeCount already is an int32, it's being upcast to int.

---

*[server/status_test.go, line 619 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOAhg93RlzLBonoQ3ef:-KOBH7G9Sema0GQP5CHl:-262907116) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec511541eed7425ff01c69f6211b28b8a376bb21/server/status_test.go#L619)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> What's up with this and the commented out code below? Why do you need to wait for IsStopped? Stop doesn't return until IsStopped is closed.
</details>
Test needed cleanup from debugging earlier issues, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[server/admin_cluster_test.go, line 141 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018#-KOBURjbY-FsqjpkTElX:-KOBURjcuS_e-JqLVBNB:1100321076) ([raw file](https://github.com/cockroachdb/cockroach/blob/9bfa125dff71de9df55581c18eef02fc95adc1f3/server/admin_cluster_test.go#L141)):*
> ```Go
> 	// detecting leaks.
> 	client.Timeout = 1 * time.Nanosecond
> 	_ = util.GetJSON(client, url, &tsResponse)
> ```

I think we still need to get to the bottom of why this is sometimes succeeding.

Perhaps we need to loop until we get a failure to guarantee that we're actually testing the not-goroutine-leak.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8018)*
<!-- Sent from Reviewable.io -->

tamird commented on util/leaktest/leaktest.go:
```diff
@@ -44,6 +44,7 @@ func interestingGoroutines() (gs []string) {
 			ignorestdlibleaks && strings.Contains(stack, "crypto/tls.(*Conn).serverHandshake") ||
 			ignorestdlibleaks && strings.Contains(stack, "net/http.(*Transport).getConn") ||
 			ignorestdlibleaks && strings.Contains(stack, "crypto/tls.(*block).readFromUntil") ||
+			ignorestdlibleaks && strings.Contains(stack, "net/http.(*Transport).dialConn") ||
```
this will need to go away on rebase

mrtracy commented:
I never got an lgtm on this; are we okay with merging?

petermattis commented:
LGTM
